File: compras-service/app/services/eventos_client.py
"""
Cliente HTTP para el servicio de Eventos (Rust)
"""
import httpx
from typing import Optional
import time
from app.config import settings
from app.utils.logger import logger, log_external_call


class EventosClient:
    """Cliente para comunicarse con el servicio de Eventos"""
    
    def __init__(self):
        self.base_url = settings.eventos_service_url
    
    async def get_evento(self, evento_id: int) -> Optional[dict]:
        """Obtiene un evento por ID"""
        endpoint = f"/eventos/{evento_id}"
        
        try:
            start_time = time.time()
            async with httpx.AsyncClient(timeout=5.0) as client:
                response = await client.get(f"{self.base_url}{endpoint}")
                duration_ms = (time.time() - start_time) * 1000
                
                log_external_call("eventos-service", endpoint, response.status_code, duration_ms)
                
                if response.status_code == 200:
                    return response.json()
                elif response.status_code == 404:
                    logger.warning(
                        f"[compras-service]: Evento no encontrado en servicio externo | "
                        f"EventoID: {evento_id}"
                    )
                    return None
                else:
                    logger.error(
                        f"[compras-service]: Error al obtener evento | "
                        f"EventoID: {evento_id} | Status: {response.status_code}"
                    )
                    return None
                    
        except httpx.RequestError as e:
            logger.error(
                f"[compras-service]: Error de conexión con eventos-service | "
                f"Endpoint: {endpoint} | Error: {str(e)}"
            )
            return None
    
    async def listar_eventos(self) -> list:
        """Lista todos los eventos disponibles"""
        endpoint = "/eventos"
        
        try:
            start_time = time.time()
            async with httpx.AsyncClient(timeout=10.0) as client:
                response = await client.get(f"{self.base_url}{endpoint}")
                duration_ms = (time.time() - start_time) * 1000
                
                log_external_call("eventos-service", endpoint, response.status_code, duration_ms)
                
                if response.status_code == 200:
                    data = response.json()
                    eventos = data.get("eventos", [])
                    logger.info(
                        f"[compras-service]: Eventos obtenidos | Total: {len(eventos)}"
                    )
                    return eventos
                else:
                    logger.error(
                        f"[compras-service]: Error al listar eventos | Status: {response.status_code}"
                    )
                    return []
                    
        except httpx.RequestError as e:
            logger.error(
                f"[compras-service]: Error de conexión con eventos-service | "
                f"Endpoint: {endpoint} | Error: {str(e)}"
            )
            return []


eventos_client = EventosClient()

File: servicio-notificaciones/src/services/eventos.client.ts
import axios from 'axios';
import config from '../config/config';
import logger, { logExternalCall } from '../utils/logger';
import { Evento } from '../models/notificacion.model';

class EventosClient {
    private baseUrl: string;

    constructor() {
        this.baseUrl = config.services.eventos;
    }

    /**
     * Obtiene datos de un evento por ID
     */
    async getEvento(eventoId: number): Promise<Evento | null> {
        const endpoint = `/eventos/${eventoId}`;
        
        try {
            const startTime = Date.now();
            const response = await axios.get(`${this.baseUrl}${endpoint}`, {
                timeout: 5000,
            });
            const durationMs = Date.now() - startTime;

            logExternalCall('eventos-service', endpoint, response.status, durationMs);

            if (response.status === 200) {
                logger.debug(
                    `[notificaciones-service]: Evento obtenido | ` +
                    `ID: ${eventoId} | Nombre: ${response.data.nombre}`
                );
                return response.data;
            }

            logger.warn(`[notificaciones-service]: Evento ${eventoId} no encontrado (status ${response.status})`);
            return null;
        } catch (error: any) {
            const durationMs = Date.now() - Date.now();
            
            if (error.response?.status === 404) {
                logExternalCall('eventos-service', endpoint, 404, durationMs);
                logger.warn(`[notificaciones-service]: Evento ${eventoId} no encontrado (404)`);
                return null;
            }
            
            logger.error(
                `[notificaciones-service]: Error al obtener evento ${eventoId}: ${error.message}`,
                { endpoint, stack: error.stack }
            );
            return null;
        }
    }
}

export default new EventosClient();